File: src/test/java/org/roborace/lapscounter/LapsCounterAbstractTest.java
package org.roborace.lapscounter;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.awaitility.Awaitility;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.roborace.lapscounter.client.WebsocketClient;
import org.roborace.lapscounter.domain.Message;
import org.roborace.lapscounter.domain.State;
import org.roborace.lapscounter.domain.Type;
import org.springframework.beans.factory.annotation.Value;

import javax.websocket.DeploymentException;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import static org.awaitility.Awaitility.await;
import static org.awaitility.Durations.FIVE_SECONDS;
import static org.awaitility.Durations.ONE_MILLISECOND;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.roborace.lapscounter.domain.Message.builder;
import static org.roborace.lapscounter.domain.State.READY;

abstract class LapsCounterAbstractTest {

    protected static final int FIRST_SERIAL = 100;
    protected static final int SECOND_SERIAL = 101;

    @Value("${local.server.port}")
    private int port = 8888;

    private static final String WS_SERVER = "ws://localhost:%d/ws";
    protected static final long TIME_SEND_INTERVAL = 10_000L;

    protected final ObjectMapper objectMapper = new ObjectMapper();


    protected WebsocketClient ui;


    @BeforeAll
    static void beforeAllAbstract() {
        Awaitility.setDefaultTimeout(FIVE_SECONDS);
        Awaitility.setDefaultPollInterval(ONE_MILLISECOND);
    }

    @BeforeEach
    void setUpAbstract() {

        ui = createClient("UI");
        givenReadyState();

        System.out.println("Given state READY");

    }

    @AfterEach
    void tearDownAbstract() {
        ui.closeClient();
    }

    protected WebsocketClient createClient(String name) {
        try {
            String scheme = String.format(WS_SERVER, port);
            return new WebsocketClient(new URI(scheme), name);
        } catch (IOException | DeploymentException | URISyntaxException e) {
            throw new RuntimeException(e);
        }
    }

    protected WebsocketClient createAndInitRobot(String name, int serial) {
        WebsocketClient robot = createClient(name);
        shouldReceiveState(robot, READY);
        sendMessage(robot, builder().type(Type.ROBOT_INIT).serial(serial).build());
        await().until(() -> ui.hasMessageWithType(Type.LAP));
        await().until(() -> robot.hasMessageWithType(Type.LAP));
        return robot;
    }

    private void givenReadyState() {
        shouldReceiveType(ui, Type.STATE);
        if (ui.getLastMessage().getState() != READY) {
            sendCommandAndCheckState(READY);
        }
    }
    protected void givenRunningState() {
        sendCommandAndCheckState(State.STEADY);
        sendCommandAndCheckState(State.RUNNING);
    }


    protected void sendCommandAndCheckState(State state) {
        sendCommand(state);
        shouldReceiveState(ui, state);
    }

    protected void sendState() {
        sendMessage(ui, buildWithType(Type.STATE));
    }

    protected void sendCommand(State state) {
        sendMessage(ui, Message.builder().type(Type.COMMAND).state(state).build());
    }

    protected void sendTimeRequestCommand(long raceTimeLimit) {
        sendMessage(ui, Message.builder().type(Type.TIME).raceTimeLimit(raceTimeLimit).build());
    }

    protected void sendMessage(WebsocketClient client, Message message) {
        try {
            client.sendMessage(objectMapper.writeValueAsString(message));
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    protected static Message buildWithType(Type type) {
        return Message.builder().type(type).build();
    }


    protected void shouldReceiveState(WebsocketClient client, State state) {
        shouldReceiveType(client, Type.STATE);
        assertThat(client.getLastMessage().getState(), equalTo(state));
    }

    protected Message shouldReceiveType(WebsocketClient client, Type type) {
        await().until(() -> client.hasMessageWithType(type));
        return client.getLastMessage();
    }

}

File: src/test/kotlin/org/roborace/lapscounter/LapsCounterAbstractTest.kt
package org.roborace.lapscounter

import com.fasterxml.jackson.databind.ObjectMapper
import org.awaitility.Awaitility
import org.awaitility.Durations
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.roborace.lapscounter.client.WebsocketClient
import org.roborace.lapscounter.domain.State
import org.roborace.lapscounter.domain.Type
import org.roborace.lapscounter.domain.api.Message
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import java.lang.Thread.sleep
import java.net.URI

internal abstract class LapsCounterAbstractTest {
    @Value("\${local.server.port}")
    private val port = 8888

    @Autowired
    private lateinit var objectMapper: ObjectMapper

    protected lateinit var ui: WebsocketClient


    @BeforeEach
    fun setUpAbstract() {
        sleep(50)
        ui = createClient("UI").also {
            givenReadyState(it)
        }

        println("Given state READY")
    }

    @AfterEach
    fun tearDownAbstract() {
        ui.closeClient()
    }

    protected fun createClient(name: String) = WebsocketClient(URI(wsServer(port)), name)

    protected fun createAndInitRobot(name: String, serial: Int) =
        createClient(name).also {
            shouldReceiveState(it, State.READY)
            sendMessage(it, Message(Type.ROBOT_INIT, serial = serial))
            Awaitility.await().until { ui.hasMessageWithType(Type.LAP) }
            Awaitility.await().until { it.hasMessageWithType(Type.LAP) }
        }

    private fun givenReadyState(client: WebsocketClient) = shouldReceiveState(client, State.READY)

    protected fun givenRunningState() {
        sendCommandAndCheckState(State.STEADY)
        sendCommandAndCheckState(State.RUNNING)
    }


    protected fun sendCommandAndCheckState(state: State) {
        sendCommand(state)
        shouldReceiveState(ui, state)
    }

    protected fun sendState() = sendMessage(ui, Message(Type.STATE))

    protected fun sendCommand(state: State) = sendMessage(ui, Message(Type.COMMAND, state = state))

    protected fun sendTimeRequestCommand(raceTimeLimit: Long) =
        sendMessage(ui, Message(Type.TIME, raceTimeLimit = raceTimeLimit))

    protected fun sendMessage(client: WebsocketClient, message: Message) =
        client.sendMessage(objectMapper.writeValueAsString(message))

    protected fun shouldReceiveState(client: WebsocketClient, state: State) {
        shouldReceiveType(client, Type.STATE) {
            assertThat(it.state, equalTo(state))
        }
    }

    protected fun shouldReceiveType(
        client: WebsocketClient,
        type: Type,
        assert: (it: Message) -> Unit = {}
    ) {
        Awaitility.await().untilAsserted {
            client.hasMessageWithType(type)
            assert(client.lastMessage)
        }
    }

    companion object {
        const val FIRST_SERIAL: Int = 100
        const val SECOND_SERIAL: Int = 101

        private fun wsServer(port: Int) = "ws://localhost:$port/ws"
        const val TIME_SEND_INTERVAL: Long = 800L

        @JvmStatic
        @BeforeAll
        fun beforeAllAbstract() {
            Awaitility.setDefaultTimeout(Durations.FIVE_SECONDS)
            Awaitility.setDefaultPollInterval(Durations.ONE_MILLISECOND)
        }
    }
}